runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/TestDataflowRunner.java

```diff
@@ -99,88 +96,115 @@ DataflowPipelineJob run(Pipeline pipeline, DataflowRunner runner) {
-            }
-            break;
-          }
-        }
+    // If there is a certain assertion failure, throw the most precise exception we can.
+    // There are situations where the metric will not be available, but as long as we recover
+    // the actionable message from the logs it is acceptable.
+    if (!allAssertionsPassed.isPresent()) {
+      LOG.warn("Dataflow job {} did not output a success or failure metric.", job.getJobId());
+    } else if (!allAssertionsPassed.get()) {
+      throw new AssertionError(errorMessage(job, messageHandler));
+    }
 
-        // Whether we canceled or not, this gets the final state of the job or times out
-        State finalState =
-            job.waitUntilFinish(
-                Duration.standardSeconds(options.getTestTimeoutSeconds()), messageHandler);
+    // Other failures, or jobs where metrics fell through for some reason, will manifest
+    // as simply job failures.
+    if (!jobSuccess) {
+      throw new RuntimeException(errorMessage(job, messageHandler));
+    }
 
-        // Getting the final state timed out; it may not indicate a failure.
-        // This cancellation may be the second
-        if (finalState == null || finalState == State.RUNNING) {
-          LOG.info(
-              "Dataflow job {} took longer than {} seconds to complete, cancelling.",
-              job.getJobId(),
-              options.getTestTimeoutSeconds());
-          job.cancel();
-        }
+    // If there is no reason to immediately fail, run the success matcher.
+    assertThat(job, testPipelineOptions.getOnSuccessMatcher());
+    return job;
+  }
 
-        if (messageHandler.hasSeenError()) {
-          result = Optional.of(false);
-        }
-      } else {
-        job.waitUntilFinish(Duration.standardSeconds(-1), messageHandler);
-        result = checkForPAssertSuccess(job);
+  /**
+   * Return {@code true} if the job succeeded or {@code false} if it terminated in any other manner.
+   */
+  private boolean waitForStreamingJobTermination(
+      final DataflowPipelineJob job, ErrorMonitorMessagesHandler messageHandler) {
+    // In streaming, there are infinite retries, so rather than timeout
+    // we try to terminate early by polling and canceling if we see
+    // an error message
+    options.getExecutorService().submit(new CancelOnError(job, messageHandler));
+
+    // Whether we canceled or not, this gets the final state of the job or times out
+    State finalState;
+    try {
+      finalState =
+          job.waitUntilFinish(
+              Duration.standardSeconds(options.getTestTimeoutSeconds()), messageHandler);
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    } catch (InterruptedException e) {
+      Thread.interrupted();
+      return false;
+    }
+
+    // Getting the final state may have timed out; it may not indicate a failure.
+    // This cancellation may be the second
+    if (finalState == null || !finalState.isTerminal()) {
```

Comment:
It is documented to do so on timeout or interruption, though it does seem like a design error.